crates/matrix-sdk/src/client/builder.rs

```diff
@@ -371,63 +402,88 @@ impl ClientBuilder {
         #[cfg(feature = "experimental-sliding-sync")]
         let mut sliding_sync_proxy: Option<Url> = None;
 
-        let homeserver = match homeserver_cfg {
-            HomeserverConfig::Url(url) => {
-                #[cfg(feature = "experimental-sliding-sync")]
-                {
-                    sliding_sync_proxy =
-                        self.sliding_sync_proxy.as_ref().map(|url| Url::parse(url)).transpose()?;
+        let (homeserver, well_known) = match homeserver_cfg {
+            HomeserverConfig::Url(url) => (url, None),
+            HomeserverConfig::ServerName { server: server_name, protocol } => {
+                let well_known = discover_homeserver(server_name, protocol, &http_client).await?;
+                (well_known.homeserver.base_url.clone(), Some(well_known))
+            }
+            HomeserverConfig::ServerNameOrUrl(server_name_or_url) => {
+                // Store the result to return at the end. If this doesn't get modified, then the
+                // supplied name is neither a server name, nor a valid URL.
+                let mut result: Result<
+                    (String, Option<discover_homeserver::Response>),
+                    ClientBuildError,
+                > = Err(ClientBuildError::InvalidServerName);
```

Comment:
It would be slightly simpler to store an `Option<(String, Option<discovery response>)>` here, set to `None` in the first place, and then only set it when it doesn't fail.

At the end of this block, if it's `None`, we can fail with `InvalidServerName`.

Reply:
If both discovery and direct URL fail, we want to return the error from discovery (as direct URL is only a fallback). So whilst I could do this, we would then also need to save the error from the well-known lookup separately. I can do that if it's a preferable pattern, but this seemed like a neater approach to me given we need both.

Reply:
I find it a bit confusing to keep a local result, and set it to an error initially, only to unset it later; even if it's slightly more code duplication (but really something we can manage), could we use the `Option` path here?

Reply:
Sure thing, will update and get this ready for a re-review :)